File: tools/cpp/naivebayes/src/naivebayes.cpp
#include <iostream>
#include <unistd.h>
#include <getopt.h>
#include <string.h>
#include <algorithm>
#include <fstream>
#include <map>
#include <vector>
#include <sstream>
#include <csv.h>

// Global vars
std::string g_inputFile;
std::string g_outputFile;
std::string g_cacheFile;
std::string g_loadCacheFile;
std::string g_sentence;

// Cache types
const std::string CACHE_WORD = "word";
const std::string CACHE_CATEGORY = "category";

/**
 * Print program usage to stdout
 */
void printUsage() {
    std::cout
            << "naivebayes - Text classification using Naive Bayes" << std::endl << std::endl
            << "Usage: naivebayes [-i input.csv] [-o output.csv]" << std::endl
            << "    -i, --input     CSV input file" << std::endl
            << "    -c, --cache     Cache calculated probabilities to a file" << std::endl
            << "    -l, --loadcache Load cache probabilities from file" << std::endl
            << "    -o, --output    Output file" << std::endl
            << "    -s, --sentence  Sentence to evaluate" << std::endl
            << "    -h, --help      Display this help message" << std::endl;
}

/**
 * Initialize parameters
 * @param argc
 * @param argv
 */
void initParams(int argc, char *argv[]) {

    struct option longOptions[] = {
            {"input", required_argument, 0, 'i'},
            {"output", required_argument, 0, 'o'},
            {"cache", required_argument, 0, 'c'},
            {"loadcache", required_argument, 0, 'l'},
            {"sentence", required_argument, 0, 's'},
            {"help",   no_argument,       0, 'h'},
            {0, 0,                        0, 0}
    };

    int optionIndex = 0;
    int c;
    while ((c = getopt_long(argc, argv, "hi:o:c:l:s:", longOptions, &optionIndex)) != -1) {
        switch (c) {
            case 'i':
                g_inputFile = optarg;
                break;
            case 'o':
                g_outputFile = optarg;
                break;
            case 'c':
                g_cacheFile = optarg;
                break;
            case 'l':
                g_loadCacheFile = optarg;
                break;
            case 's':
                g_sentence = optarg;
                break;
            case 'h':
            default:
                break;
        }
    }
}

/**
 * Convert a string to a vector of tokens
 * @param text
 * @return vector of words
 */
std::vector<std::string> tokenize(std::string text) {
    std::vector<std::string> tokens;
    char *cstr = new char[text.length() + 1];
    strcpy(cstr, text.c_str());
    char * pch;
    pch = strtok (cstr," \t");
    while (pch != NULL) {
        tokens.push_back(pch);
        pch = strtok (NULL, " ,.-");
    }
    delete [] cstr;
    return tokens;
}

int main(int argc, char** argv) {

    // Initialize parameters
    initParams(argc, argv);

    // Handle errors
    if((g_loadCacheFile.empty() && g_inputFile.empty()) || g_outputFile.empty()) {
        printUsage();
        return 0;
    }

    try{

        // Naive Bayes variables
        std::map<int, unsigned long> categoryCounter;
        std::map<std::string, std::map<int, unsigned long>> wordGivenCategoryCounter;
        std::map<int, unsigned long> wordsInCategoryCounter;
        std::map<std::string, std::map<int, double>> pWordGivenCategory;
        std::map<int, double> pCategory;
        unsigned long totalUtt = 0;

        if(g_loadCacheFile.empty()) {

            // Prepare csv variables
            int id;
            int category;
            std::string text;

            // Load information
            std::cout << ">> Loading: " << g_inputFile << std::endl;
            io::CSVReader<3> inputCSV(g_inputFile);
            inputCSV.read_header(io::ignore_extra_column, "Id", "Category", "Text");
            while(inputCSV.read_row(id, category, text)){

                // Convert string to tokens
                std::vector<std::string> tokens = tokenize(text);
                for(std::string word : tokens) {

                    // The number of times a word occurs in a category
                    wordGivenCategoryCounter[word][category]++;
                }

                // The number of utterances in this category
                categoryCounter[category]++;

                // Increment the total number of utterance
                totalUtt++;

                // THe number of words in this category
                wordsInCategoryCounter[category] += tokens.size();
            }

            // Start calculating the probabilities
            const int BIAS = 1;
            std::cout << ">> Computing P(Wi|Cj)" << std::endl;
            for(auto &wordEntry : wordGivenCategoryCounter) {
                for(auto &categoryEntry : categoryCounter) {
                    pWordGivenCategory[wordEntry.first][categoryEntry.first] = (double)
                                                                                       (BIAS +  wordGivenCategoryCounter[wordEntry.first][categoryEntry.first]) /
                                                                               (wordsInCategoryCounter[categoryEntry.first] + wordGivenCategoryCounter.size());
                }
            }

            std::cout << ">> Computing P(Cj)" << std::endl;
            for(auto entry : categoryCounter) {
                pCategory[entry.first] = (double) entry.second / totalUtt;
            }
        } else {
            // Prepare csv variables
            std::string word;
            int category;
            std::string type;
            double probability;

            // Load information
            std::cout << ">> Loading probabilities from cache: "<< g_loadCacheFile << std::endl;
            io::CSVReader<4> cacheCSV(g_loadCacheFile);
            cacheCSV.read_header(io::ignore_extra_column, "Word", "Category", "Probability", "Type");
            while(cacheCSV.read_row(word, category, probability, type)) {
                if(type == CACHE_CATEGORY) {
                    pCategory[category] = probability;
                } else if(type == CACHE_WORD) {
                    pWordGivenCategory[word][category] = probability;
                } else {
                    std::cerr << "Unknown type " << type << std::endl;
                }
            }
        }

        // Cache results if requested
        if(!g_cacheFile.empty()) {
            std::cout << ">> Caching computed probabilities at: " << g_cacheFile << std::endl;
            std::ofstream cacheFile(g_cacheFile);

            // If cache file was opened
            if(cacheFile.is_open()) {
                cacheFile << "Word,Category,Probability,Type" << std::endl;

                // Cache probabilities for the categories
                for(auto entry : pCategory) {
                    cacheFile << "," << entry.first << "," << entry.second << "," << CACHE_CATEGORY << std::endl;
                }

                // Cache probabilities for words given categories
                for(auto wordEntry : pWordGivenCategory) {
                    for(auto categoryEntry : wordEntry.second) {
                        cacheFile << wordEntry.first << "," << categoryEntry.first << ","
                                  << categoryEntry.second << "," << CACHE_WORD << std::endl;
                    }
                }
                cacheFile.close();
            } else {
                std::cerr << "Could not open cache file: " << g_cacheFile << std::endl;
            }
        }

        // Evaluate a sentence
        if(!g_sentence.empty()) {
            std::cout << ">> Evaluating sentence: " << g_sentence << std::endl;

            // Convert sentence to tokens
            std::vector<std::string> tokens = tokenize(g_sentence);
            for(auto entry : pCategory) {
                double val = entry.second;
                for(std::string token : tokens) {
                    if(pWordGivenCategory.find(token) != pWordGivenCategory.end()) {
                        val *= pWordGivenCategory[token][entry.first];
                    }
                }
                std::cout << "P(S | " << entry.first << ") = " << val << std::endl;
            }
        }

    } catch (io::error::can_not_open_file exception) {
        std::cerr << "Error opening input file" << std::endl;
        return 1;
    }
    return 0;
}

File: tools/python/naivebayes/naivebayes.py
import argparse
import csv

class CLI:
    def read(self):
        """Initialize a command line interface"""

        # Define arguments
        parser = argparse.ArgumentParser(description='Analyze csv file and apply naive bayes')
        parser.add_argument('-i','--input', nargs=1, help='CSV input file')
        parser.add_argument('-c','--cache', nargs=1, help='Cache calculated probabilities to a file')
        parser.add_argument('-l','--loadcache', nargs=1, help='Load cached probabilities from file')
        parser.add_argument('-s','--sentence', nargs=1, help='Sentence to evaluate')
        parser.add_argument('-t','--test', nargs=1, help='Load test sentences from CSV file')
        parser.add_argument('-o','--out', nargs=1, help='Output test result to CSV file or -- for stdout')
        args = parser.parse_args()

        # Create a naive bayes instance
        naiveBayes = NaiveBayes()

        # Checkfor missing arguments
        if args.input is None and args.loadcache is None or args.out is None:
            print("Missing arguments")
            exit(1)

        # Load input csv
        if args.input is not None:
            print(">> Loading:", args.input[0])
            with open(args.input[0], newline='', encoding='utf-8') as csvfile:
                reader = csv.DictReader(csvfile)
                naiveBayes.m_totalUtt = len(reader)
                for row in reader:
                    # Load attributes
                    id = row['Id']
                    category = row['Category']
                    text = row['Text']

                    # Tokenize text
                    tokens = text.split()
                    for token in tokens:

                        # Init first dimension
                        if token not in naiveBayes.m_wordGivenCategoryCounter:
                            naiveBayes.m_wordGivenCategoryCounter[token] = {}

                        # Init second dimension
                        if category not in naiveBayes.m_wordGivenCategoryCounter[token]:
                            naiveBayes.m_wordGivenCategoryCounter[token][category] = 0

                        # Increment counter
                        naiveBayes.m_wordGivenCategoryCounter[token][category] += 1


                    # Init category counter
                    if category not in naiveBayes.m_categoryCounter:
                        naiveBayes.m_categoryCounter[category] = 0

                    # Increment category counter
                    naiveBayes.m_categoryCounter[category] += 1

                    # Init words per category counter
                    if category not in naiveBayes.m_wordsInCategoryCounter:
                        naiveBayes.m_wordsInCategoryCounter[category] = 0

                    # Increment words in category
                    naiveBayes.m_wordsInCategoryCounter[category] += len(tokens)
            
            # Compute probabilities
            naiveBayes.compute()

class NaiveBayes:
    def __init__(self):
        """Initialize naive bayes variables"""
        
        # Store counts
        self.m_totalUtt = 0
        self.m_categoryCounter= {}
        self.m_wordGivenCategoryCounter = {}
        self.m_wordsInCategoryCounter = {}

        # Store probabilities
        self.m_pWordGivenCategory = {}
        self.m_pCategory = {}

    def compute(self):
        """Compute probabilities"""

        BIAS = 1
        print(">> Computing P(Wi|Cj)")
        for word in self.m_wordGivenCategoryCounter:

            # Init word probability
            if word not in self.m_pWordGivenCategory:
                self.m_pWordGivenCategory[word] = {}

            # Compute probabilities
            for category in self.m_categoryCounter:
                if category in self.m_wordGivenCategoryCounter[word]:
                    self.m_pWordGivenCategory[word][category] = \
                            (BIAS +  self.m_wordGivenCategoryCounter[word][category]) / \
                            (self.m_wordsInCategoryCounter[category] + wordGivenCategoryCounter.size())


# Stat application
cli = CLI()
cli.read()
